### Use mtcars data set(built-in in R)

a. Run linear regression of mpg on hp, explain the coefficient(in real life terms, with the unit of variable) and goodness of fit.

b. Run a polynomials regression of mpg on hp, up to 6 degree, use the rank matrix to select the model you think is the best in terms of bias-variance trade-off, justify your answers.

In [1]:
 #head(mtcars)
 #?mtcars
 fit= lm(mtcars$mpg~mtcars$hp,data = mtcars)
 summary(fit)


Call:
lm(formula = mtcars$mpg ~ mtcars$hp, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.7121 -2.1122 -0.8854  1.5819  8.2360 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 30.09886    1.63392  18.421  < 2e-16 ***
mtcars$hp   -0.06823    0.01012  -6.742 1.79e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.863 on 30 degrees of freedom
Multiple R-squared:  0.6024,	Adjusted R-squared:  0.5892 
F-statistic: 45.46 on 1 and 30 DF,  p-value: 1.788e-07


From the model, we can say that coefficient is statistically significant enough to explain the relationship between mileage(mpg) and power of engine(hp).
 
Since the F-statistic is much larger than 1 and has a correspondingly small p-value we can confirm the likelihood that the null hypothesis is false and that there is a statistically significant relationship between mpg and horsepower.
 
The linear equation for the model can be written as:

#### mpg = 30.09 + (-0.068) * hp

 Looking at the sign of the regression coefficient of horsepower,we see that the relationship between mpg and horsepower is negative.
 In simple language, with increase in one unit of horsepower for a engine, the mileage decreases by 0.068. In other words, The distancein miles an engine/car can take you from one gallon of fuel decreses by 0.068 with one horsepower increase in engine.
 
 The goodness of fit, given by R-squared = 0.6024 indicates that 60.24% of the variance of mpg is explained by horsepower. This means that the relationship between mpg and horsepower is strong.

In [2]:
 model_rank = data.frame() #Defining an object to store rank,etc
 fit= lm(mtcars$mpg~mtcars$hp,data = mtcars) #Degree-1 Regression
 summary(fit)


Call:
lm(formula = mtcars$mpg ~ mtcars$hp, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.7121 -2.1122 -0.8854  1.5819  8.2360 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 30.09886    1.63392  18.421  < 2e-16 ***
mtcars$hp   -0.06823    0.01012  -6.742 1.79e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.863 on 30 degrees of freedom
Multiple R-squared:  0.6024,	Adjusted R-squared:  0.5892 
F-statistic: 45.46 on 1 and 30 DF,  p-value: 1.788e-07


In [3]:
rsq = summary(fit)$r.squared
 t=1
 f=0
 model_details = c(1,rsq,t,f,(t/1))
 model_rank = rbind(model_rank,model_details)
 fit= lm(mtcars$mpg~poly(mtcars$hp,2),data = mtcars) #Degree-2 Regression
 summary(fit)


Call:
lm(formula = mtcars$mpg ~ poly(mtcars$hp, 2), data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.5512 -1.6027 -0.6977  1.5509  8.7213 

Coefficients:
                    Estimate Std. Error t value Pr(>|t|)    
(Intercept)           20.091      0.544  36.931  < 2e-16 ***
poly(mtcars$hp, 2)1  -26.046      3.077  -8.464 2.51e-09 ***
poly(mtcars$hp, 2)2   13.155      3.077   4.275 0.000189 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.077 on 29 degrees of freedom
Multiple R-squared:  0.7561,	Adjusted R-squared:  0.7393 
F-statistic: 44.95 on 2 and 29 DF,  p-value: 1.301e-09


In [4]:
 rsq = summary(fit)$r.squared
 t=2
 f=0
 model_details = c(2,rsq,t,f,(t/2))
 model_rank = rbind(model_rank,model_details)
 for(i in 3:6){
 fit= lm(mtcars$mpg~poly(mtcars$hp,i),data = mtcars) #Loop for Degree-3 to 6 Regression
 print(summary(fit))
 rsq = summary(fit)$r.squared
 t = table((summary(fit)$coefficients[,4])<0.05)[2]-1
 f = table((summary(fit)$coefficients[,4])<0.05)[1]
 model_details = c(i,rsq,t,f,(t/i))
 model_rank = rbind(model_rank,model_details)
 }


Call:
lm(formula = mtcars$mpg ~ poly(mtcars$hp, i), data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.8605 -1.3972 -0.5736  1.6461  9.0738 

Coefficients:
                    Estimate Std. Error t value Pr(>|t|)    
(Intercept)          20.0906     0.5485  36.625  < 2e-16 ***
poly(mtcars$hp, i)1 -26.0456     3.1030  -8.394 3.95e-09 ***
poly(mtcars$hp, i)2  13.1546     3.1030   4.239 0.000221 ***
poly(mtcars$hp, i)3  -2.2419     3.1030  -0.722 0.475987    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.103 on 28 degrees of freedom
Multiple R-squared:  0.7606,	Adjusted R-squared:  0.7349 
F-statistic: 29.65 on 3 and 28 DF,  p-value: 7.769e-09


Call:
lm(formula = mtcars$mpg ~ poly(mtcars$hp, i), data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.8914 -1.3842 -0.5696  1.6120  9.1050 

Coefficients:
                    Estimate Std. Error t value Pr(>|t|)    
(Intercept)          20.0906     0.5585  3

In [7]:
colnames(model_rank) = c("model_order","R-Squared","Coef_true","Coef_false","Goodness")
Rank = sqrt((model_rank[,2])^2 + (model_rank[,5])^2)
model_rank = cbind(model_rank,Rank)
model_rank #Final table to observe Rank for each model

model_order R-Squared Coef_true Coef_false Goodness  Rank     
1 1           0.6024373 1         0          1.0000000 1.1674463
2 2           0.7561100 2         0          1.0000000 1.2536755
3 3           0.7605735 2         1          0.6666667 1.0113933
4 4           0.7606511 2         2          0.5000000 0.9102692
5 5           0.7622443 2         3          0.4000000 0.8608231
6 6           0.7819456 2         4          0.3333333 0.8500295

 The best model for our data should be Polynomial Regression with Degree-2. It has the highest value of Rank. The rank has been calculated as:
 
 #### Rank = sqrt [(R-Squared)ˆ2 + (Goodness)ˆ2]
 
 where, Goodness = Number of statistically significant coefficients in a model/ Polynomial Order The bias is an error from erroneous assumptions in the learning algorithm. High bias (overly simple) can cause an algorithm to miss the relevant relations between features and target outputs (underfitting).
 
 The variance is an error from sensitivity to small fluctuations in the training set. High variance (overly complex) can cause an algorithm to model the random noise in the training data, rather than the intended outputs (overfitting). A model can be overfit if it is too specific to the data presented and offers little explanatory power for any new data that might come into the system. The more complex your model is(higher degree of polynomial in this case), the tighter it’s going to fit to your data, but we run the risk of some of the coefficients being less statistically valid as the model becomes more complex.
 
 The result table shows that, although the model’s R-squared accuracy is increasing as the fit becomes more complex(R-square is maximum for degree-6), but the goodness of that fit is going down with increasing order of models. This is because the number of statistically significant coefficients compared to the total number
 of coefficients in a model tends to go down with incresing model orders. We should neither select a model with maximum R-Squared value nor the one with maximum Goodness. Selecting model with maximum R-Squared value will give very accurate model but it will be very complex model with many insignificant coefficients. On the other hand, we should also not select the model with maximum goodness of fit, because it will reduce the complexity but will make model inaccurate.
 
 We can see where the trade-off is best between model accuracy and goodness of fit.This is why we are using Rank as indicator of best model.
 The model order with the highest rank in this case is a Degree-2 fit that has all of its coefficients that are statistically valid and also decent R-squared value.
 Wechoose Optimal model to be the one that has the highest rank value. A model that has a lower R-squared and lower rank is underfit to the data. A model that has a higher R-squared and a lower rank is overfit to our data.